### ISSUE TITLE
Invalid C code for inline iterator with inner proc

### ISSUE THREAD

**Issue by DomoSokrat** [nim-lang/Nim#14165]
While trying out the different nim solutions to the Hamming numbers task from RosettaCode I found that one variant did not compile. Reducing it to the max gave me this:

### 
```nim
iterator log_nodups_hamming(): int {.inline.} =
  let lb3 = 1
  proc mul3(): int = lb3
  yield mul3()

for h in log_nodups_hamming():
  break
for h in log_nodups_hamming():
  break
```

### Output
```
/mnt/space/data/nim/.cache/hn_5_test/d/@mhn_5_test.nim.c: In function ‘NimMainModule’:
/mnt/space/data/nim/.cache/hn_5_test/d/@mhn_5_test.nim.c:214:5: error: ‘colonenv_’ undeclared (first use in this function); did you mean ‘colonenv__2’?
  214 |   (*colonenv_).lb31 = ((NI) 1);
      |     ^~~~~~~~~
      |     colonenv__2
/mnt/space/data/nim/.cache/hn_5_test/d/@mhn_5_test.nim.c:214:5: note: each undeclared identifier is reported only once for each function it appears in
```
This is the expansion of the second loop.

### Additional Information

* Original source: https://rosettacode.org/wiki/Hamming_numbers#Imperative_iterator_implementation_of_the_above_functional_version
* Using `--gc:arc` the compilation succeeds.

#### Used versions
```
$ nim -v
Nim Compiler Version 1.2.0 [Linux: amd64]
Compiled at 2020-04-03
Copyright (c) 2006-2020 by Andreas Rumpf

active boot switches: -d:release
```

* The 1.2.1 nightly from 28.4. and the 1.3.1 nightly from 29.4. show the same problem.

**Comment by alaviss** [comment]
Regression happened between 0.19.0 and 0.20.0. I did two bisect run and got different results:

- 2fecf4f36a39049e2d21065929f44dff38cf7d5a
- b1ff37c08f80b440e5ca0e4281a8967396efe7a5

**Comment by ghost** [comment]
@timotheecour your https://github.com/nim-lang/Nim/issues/18536 seems to be a duplicate of this issue, can you please decide which one of the two to close?

**Comment by timotheecour** [comment]
ok, closing the other one in favor of this one; there is some complementary information so I'm pasting the issue here:

-----------------------
reduced from https://github.com/nim-lang/Nim/pull/18535#issuecomment-882957657 /cc @brightly-salty 

### Example
```nim
# iterator envPairs*(): int {.closure.} = # Error: internal error: environment misses: foo
iterator envPairs*(): int =
  var foo: seq[int]
  proc fun() =
    foo = @[]
  fun()
  yield 3

proc main() =
  for a in envPairs():
    discard
  for a in envPairs():
    discard
static: main() # nim c in vm: ok
# static: main() # nim js in vm: Error: cannot evaluate at compile time: foo
main() # nim c: @mt12609.nim.c:372:5: error: use of undeclared identifier 'colonenv_'; did you mean 'colonenv__2'?
# main() # nim js: # Error: internal error: symbol has no generated name: foo
```

### Current Output
```
/@mt12609.nim.c:372:5: error: use of undeclared identifier 'colonenv_'; did you mean 'colonenv__2'?
                (*colonenv_).foo1 = (tySequence__qwqHTkRvwhrRyENtudHQ7g*)0;
                  ^~~~~~~~~
                  colonenv__2
/Users/timothee/git_clone/nim/timn/build/nimcache/@mt12609.nim.c:364:65: note: 'colonenv__2' declared here
                tyObject_Env_t12609dotnim_envPairs___Dwg4Sd9awieN34NfR6DSgaw* colonenv__2;
                                                                              ^
/Users/timothee/git_clone/nim/timn/build/nimcache/@mt12609.nim.c:375:46: error: use of undeclared identifier 'colonenv_'; did you mean 'colonenv__2'?
                T4_.ClP_0 = fun_t4950544857_3; T4_.ClE_0 = colonenv_;

```
### Expected Output
works

### Additional Information
1.5.1 73ce40aaf71196faa08072216bd826a94b5d40b1

**Comment by Cloudperry** [comment]
I'm not sure, but I think I ran into this same bug using an anonymous proc passed to another proc. The error message is exactly the same except for the names. I will post an example here if I manage to get it minimized, but its hard because my bug happened with regex-nim and removing regex-nim usage in my program makes it compile. 

By the way the original example posted here now doesn't compile with arc on Nim 1.4.8.

**Comment by ringabout** [comment]
Hi @Cloudperry, could you post your example?

**Comment by Cloudperry** [comment]
Unfortunately it happened in one of my super small projects and I didn't have the code in any version control system. I got rid of that code because I thought that an example with regex-nim wouldn't be that useful. I tried to make the same program again from what I remember but I didn't manage to reproduce it.
